### PR TITLE
Issue #15078: Add new input file for sealed classes in ClassFanOutCom…

### LINKED ISSUE #15078: Add Check Support for Java 17 Sealed Classes: ClassFanOutComplexitiy
child of https://github.com/checkstyle/checkstyle/issues/14969

Check documentation: https://checkstyle.org/checks/metrics/classfanoutcomplexity.html

---

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="ClassFanOutComplexity">
            <property name="max" value="0"/>
        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
public class Test implements I, J {  // violation
}

sealed class C permits D {   // should this be a violation ?
}
final class D extends C {   // violation
}
interface I {
}
interface J {
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:1:1: Class Fan-Out Complexity is 2 (max allowed is 0). [ClassFanOutComplexity]
[ERROR] D:\CS\test\src\Test.java:6:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 2 errors.
PS D:\CS\test> 
```

---

from the doc: 

> Checks the number of other types a given class/record/interface/enum/annotation relies on. Also, the square of this has been shown to indicate the amount of maintenance required in functional programs (on a file basis) at least.

I don't think we should add 1 for the classes in the `permits` list. The super class `C` permitted certain classes to extend it but it doesn't rely on the permitted subclasses. the opposite is correct

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #15078

### CODE REVIEW

**Comment by mahfouz72** [comment]
> and I think documentation has to be added since we declared in documentation how we process things.

This is blocked because the example test is still disabled for this check. Also, I think it is pretty obvious from the check description that the check will not count the classes in the permits list (because the base class doesn't depend on those classes). we can keep doc improvements in another PR when we enable examples test. what do you think?

**Comment by nrmancuso** [comment]
Having the example test disabled doesn't block anything in this PR, please update the docs here

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexitySealedClasses.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.metrics.classfanoutcomplexity;
+
+// violation below, 'Class Fan-Out Complexity is 2 (max allowed is 0)'
+public class InputClassFanOutComplexitySealedClasses implements I, J {
+}
+
+// ok below, 'sealed classes don't rely on the permitted subclasses'
```
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexitySealedClasses.java:28]: Are we forced to have quotes around `'sealed classes don't rely on the permitted subclasses'`? I find this strange.

**Comment by mahfouz72** [comment]
github, generate site

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexitySealedClasses.java (hunk 1 above)
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexitySealedClasses.java]: Removed.

**Review by nrmancuso — APPROVED** [review]

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/issues/15078#issuecomment-2184315859
>  I think documentation has to be added since we declared in documentation how we process things.

I am not seeing this. Are we not doing this?

**Comment by nrmancuso** [comment]
> [#15078 (comment)](https://github.com/checkstyle/checkstyle/issues/15078#issuecomment-2184315859)
> 
> > I think documentation has to be added since we declared in documentation how we process things.
> 
> I am not seeing this. Are we not doing this because this is an input only?

The example test with sealed classes was updated, I am good with this.

**Comment by mahfouz72** [comment]
> I am not seeing this. Are we not doing this because this is an input only?

Examples were updated to have sealed in it. Isn't this what you asked for?

**Review by rnveach — APPROVED** [review]

**Comment by romani** [comment]
Github, generate report by config from InputClassFanOutComplexitySealedClasses.java

**Comment by romani** [comment]
This is expected error

**Pull request merged**